keps/sig-network/2030-endpointslice-subsetting/README.md

```diff
@@ -0,0 +1,217 @@
+topologies. A simple pattern like this will allow EndpointSlices to be delivered
+to consumers in a specific zone or region.
+
+
+### Risks and Mitigations
+
+This approach does not allow a single EndpointSlice to target multiple zones or
+regions. Any approach that enabled that would be significantly more complicated.
+The initial proposal in [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004) suggests that
+this won't be an issue. 
+
+## Design Details
+
+### API Consumption
+
+EndpointSlice API consumers can use the following selectors to consume
+EndpointSlices:
+
+#### Select EndpointSlices Without Labels
+This is required to support producers of EndpointSlices that don't set these
+labels, including older versions of the EndpointSlice controller.
+
+```
+matchExpressions:
+  - {key: endpointslice.kubernetes.io/for-zone, operator: DoesNotExist}
+  - {key: endpointslice.kubernetes.io/for-region, operator: DoesNotExist}
+```
+
+#### Select EndpointSlices With Matching Zone
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-zone: example-zone
+```
+
+#### Select EndpointSlices With Matching Region
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-region: example-region
+```
+
+#### Kube-Proxy
+When the `EndpointSliceSubsetting` feature gate is set to true on `Kube-Proxy`,
+it will use these selectors to filter EndpointSlices.
+
+### Test Plan
+This KEP is quite small in scope. The only new functionality being added will be
+an adjustment to the EndpointSlices kube-proxy consumes when a feature gate is
+enabled. We will need to add more test coverage for when this feature is enabled
+or disabled.
+
+#### Unit Tests
+* Ensure kube-proxy will continue to consume all EndpointSlices when this
+  feature is disabled.
+* Ensure EndpointSlices delivered to a specific zone will be consumed by
+  kube-proxy running in the same zone when this feature is enabled.
+* Ensure EndpointSlices delivered to a specific zone will not be consumed by
+  kube-proxy running in a different zone when this feature is enabled.
+* Ensure EndpointSlices delivered to a specific region will be consumed by
+  kube-proxy running in the same region when this feature is enabled.
+* Ensure EndpointSlices delivered to a specific region will not be consumed by
+  kube-proxy running in a different region when this feature is enabled.
+
+### Graduation Criteria
+
+#### Alpha Release
+
+- Proposed labels are added as well known labels in Discovery API types.
+- Implement new selectors in kube-proxy.
+- Implement test plan.
+
+#### Alpha -> Beta Graduation
+
+- EndpointSlice controller supports publishing EndpointSlices in subsets. (See
+  [KEP 2004](https://github.com/kubernetes/enhancements/issues/2004) for more
+  info).
+
+### Upgrade / Downgrade Strategy
+
+This will be an opt-in feature on kube-proxy. Enabling or disabling the
+`EndpointSliceSubsetting` feature gate will be all that is needed to turn this
+feature on or off.
```

Comment:
I though that feature-gates were not meant to be used as opt-in option, and the goal is to eventually remove them and make the feature enabled by default, but I' m sure @liggitt will know better.
If that' s the case this should be a configuration option for kube-proxy

Reply:
they are not intended as a long-term config option

From https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior:

> Feature gates are intended to cover the development life cycle of a feature - they are not intended to be long-term APIs. As such, they are expected to be deprecated and removed after a feature becomes GA or is dropped.

If this is expected to remain configurable / opt-in even after graduation, it should have a dedicated option.

Ideally, the design and feature would be structured so that it does not require configuration and can behave consistently on all clusters after graduation.

Reply:
Thanks for the feedback! I need to reword this. This was intended to be used like any other feature gate - guarding functionality, not a long term toggle.

Reply:
Updated this section to be more clear.